typing/printtyp.ml

```diff
@@ -1658,55 +1660,84 @@ let hide_rec_items ids =
+  { pre_ghosts: Types.signature_item list; group:rec_item_group }
 
 let group_syntactic_items x =
-  let rec group ~ghosts ~acc = function
-    | Sig_class _ as src :: ctydecl :: tydecl1 :: tydecl2 :: rem ->
-        let s_elt = {src; ghosts=ghosts @ [ctydecl; tydecl1; tydecl2]} in
-        remove_row (s_elt :: acc) rem
-    | Sig_class_type _ as src :: tydecl1 :: tydecl2 :: rem ->
-        remove_row ({src; ghosts = ghosts @ [tydecl1; tydecl2]}::acc) rem
-    | [Sig_class _] | [Sig_class _; _ ] | [Sig_class _; _;_ ] -> assert false
-    | [Sig_class_type _] | [Sig_class_type _; _ ] -> assert false
+  let rec group ~acc = function
+    | Sig_class _ as src :: rem ->
+       let ctydecl, tydecl1, tydecl2, rem =
+         match rem with
+         | cty :: tydecl1 :: tydecl2 :: rem -> cty, tydecl1, tydecl2, rem
+         | _ ->  (* a class declaration for [c] is followed by the ghost
+                    declarations of class type [c], and types [c] and [#c] *)
+            assert false
+       in
+        let s_elt =
+          { src; post_ghosts= [ctydecl; tydecl1; tydecl2]}
+        in
+        group ~acc:(s_elt :: acc) rem
+    | Sig_class_type _ as src :: rem ->
+       let tydecl1, tydecl2, rem =
+         match rem with
+         | tydecl1 :: tydecl2 :: rem -> tydecl1, tydecl2, rem
+         | _ ->  (* a class type declaration for [ct] is followed by the ghost
+                    declarations of types [ct] and [#ct] *)
+            assert false
+       in
+       group
+         ~acc:({src; post_ghosts = [tydecl1; tydecl2]}::acc)
+         rem
     | (Sig_module _ | Sig_value _ | Sig_type _ | Sig_typext _
       | Sig_modtype _ as src)  :: rem ->
-        remove_row ({src; ghosts} :: acc) rem
-    | [] -> List.rev acc
-  and remove_row acc = function
-    | Sig_type(id,_,_,_) as row :: rest when is_row_name (Ident.name id) ->
-        group ~ghosts:[row] ~acc rest
-    | rest -> group ~ghosts:[] ~acc rest in
-  remove_row [] x
+        group ~acc:({src; post_ghosts=[]} :: acc) rem
+    | [] -> List.rev acc in
+  group  ~acc:[] x
 
 let add_sigitem env x =
-  Env.add_signature (x.src :: x.ghosts) env
+  Env.add_signature (x.src :: x.post_ghosts) env
 
 let recursive_sigitem = function
-  | Sig_type(id, _, rs, _) -> Some((true,id),rs)
-  | Sig_class(id,_,rs,_) | Sig_class_type (id,_,rs,_)
-  | Sig_module(id, _, _, rs, _) -> Some ((false,id),rs)
+  | Sig_type(ident, _, rs, _) -> Some({hide=true;ident},rs)
+  | Sig_class(ident,_,rs,_) | Sig_class_type (ident,_,rs,_)
+  | Sig_module(ident, _, _, rs, _) -> Some ({hide=false;ident},rs)
   | Sig_value _ | Sig_modtype _ | Sig_typext _  -> None
 
 let group_recursive_items x =
-  let rec_group ids group = Rec_group( List.rev ids, List.rev group) in
-  let rec not_in_group acc = function
+  let rec_group pre ids group =
+    let group = Rec_group(List.rev ids, List.rev group) in
+    { pre_ghosts=List.rev pre; group } in
+  let rec not_in_group ~pre acc = function
   | [] -> List.rev acc
```

Comment:
Here there is a risk of losing information if `pre` is not empty. Maybe you could `assert (pre = [])`?

Reply:
Good idea.